### PR TITLE
Watch for the creation of missing files

### PR DESCRIPTION
**Pull request by amcasey**
When we discover that a file is missing, call `watchFile` on its path so that we can update our state when it is created.

1) For `tsc --watch`, we schedule a new build
2) For `tsserver`, we update the containing projects

Caveat: VS will not know about the state change in tsserver until the next time it synchronizes its project list

### CODE REVIEW

**Comment by amcasey** [comment]
This PR supersedes https://github.com/Microsoft/TypeScript/pull/16527.

File: src/compiler/types.ts [hunk 1]
```diff
@@ -2427,6 +2427,12 @@ namespace ts {
         getSourceFiles(): SourceFile[];
 
         /**
+         * Get a list of file names that were passed to 'createProgram' or referenced in a
+         * program source file but could not be located.
+         */
+        getMissingFilePaths?(): Path[];
```
Comment (amcasey) [inline src/compiler/types.ts]: This doesn't need to be optional.

File: src/compiler/program.ts [hunk 2]
```diff
@@ -862,6 +865,24 @@ namespace ts {
...
             }
 
+            // If a file has ceased to be missing, then we need to discard some of the old
+            // structure in order to pick it up.
+            // Caution: if the file has created and then deleted between since it was discovered to
+            // be missing, then the corresponding file watcher will have been closed and no new one
+            // will be created until we encounter a change that prevents complete structure reuse.
+            // During this interval, creation of the file will go unnoticed.  We expect this to be
+            // both rare and low-impact.
+            const missingFilePaths: Path[] = oldProgram.getMissingFilePaths() || emptyArray;
```
Comment (andy-ms) [inline src/compiler/program.ts]: It looks like `getMissingFilePaths` already never returns `undefined`. (Let's keep it that way.) Inline this declaration.

File: src/compiler/program.ts [hunk 3]
```diff
@@ -862,6 +865,24 @@ namespace ts {
...
+            for (const missingFilePath of missingFilePaths) {
+                if (host.fileExists(missingFilePath)) {
+                    return oldProgram.structureIsReused = StructureIsReused.SafeModules;
+                }
+            }
+
+            for (const p of oldProgram.getMissingFilePaths()) {
+                filesByName.set(p, undefined);
```
Comment (andy-ms) [inline src/compiler/program.ts]: The type of `filesByName` should indicate that this can be undefined.

File: src/compiler/sys.ts [hunk 4]
```diff
@@ -340,11 +340,20 @@ namespace ts {
                     }
 
                     function fileChanged(curr: any, prev: any) {
```
Comment (andy-ms) [inline src/compiler/sys.ts]: It was this way before, but it would be nice to remove `any` and include a type declaration saying what you expect this to be.

File: src/compiler/tsc.ts [hunk 5]
```diff
@@ -285,6 +285,17 @@ namespace ts {
 
             setCachedProgram(compileResult.program);
             reportWatchDiagnostic(createCompilerDiagnostic(Diagnostics.Compilation_complete_Watching_for_file_changes));
+
+            const missingPaths = compileResult.program.getMissingFilePaths() || [];
+            missingPaths.forEach((path: Path): void => {
+                const fileWatcher = sys.watchFile(path, (_fileName: string, removed?: boolean) => {
+                    // removed = deleted ? true : (added ? false : undefined)
```
Comment (andy-ms) [inline src/compiler/tsc.ts]: This could use an enum rather than `true | false | undefined`. `true` -> `E.Deleted`, `false` -> `E.Added`, `undefined` -> `E.Changed`.

File: src/compiler/types.ts [hunk 6]
```diff
@@ -2427,6 +2427,12 @@ namespace ts {
         getSourceFiles(): SourceFile[];
 
         /**
+         * Get a list of file names that were passed to 'createProgram' or referenced in a
+         * program source file but could not be located.
+         */
+        getMissingFilePaths(): Path[];
```
Comment (andy-ms) [inline src/compiler/types.ts]: Would make this internal. (`/* @internal */`)

File: src/harness/unittests/programMissingFiles.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            content: ""
+        };
+
+        const referenceFileName = "reference.ts";
+        const referenceFileRelativePath = "./" + referenceFileName;
+
+        const referenceFile: Harness.Compiler.TestFile = {
+            unitName: referenceFileName,
+            content:
+                "/// <reference path=\"d:/imaginary/nonexistent1.ts\"/>\n" + // Absolute
```
Comment (andy-ms) [inline src/harness/unittests/programMissingFiles.ts]: Is it necessary to test all the different kinds of paths here? Also, would include a reference to an existing file for contrast.

File: src/harness/unittests/programMissingFiles.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            const program = createProgram([emptyFileRelativePath], options, testCompilerHost);
+            const missing = program.getMissingFilePaths();
+            assert.isDefined(missing);
+            assert.equal(missing.length, 0);
+        });
+
+        it("handles missing root file", () => {
+            const program = createProgram(["./nonexistent.ts"], options, testCompilerHost);
+            const missing = program.getMissingFilePaths();
+            assert.isDefined(missing);
```
Comment (andy-ms) [inline src/harness/unittests/programMissingFiles.ts]: Use `assert.deepEqual` instead of asserting the length and elements separately.

File: src/server/project.ts [hunk 9]
```diff
@@ -107,6 +107,7 @@ namespace ts.server {
         private rootFilesMap: FileMap<ScriptInfo> = createFileMap<ScriptInfo>();
         private program: ts.Program;
         private externalFiles: SortedReadonlyArray<string>;
+        private missingFilesMap: FileMap<FileWatcher> = createFileMap<FileWatcher>();
```
Comment (andy-ms) [inline src/server/project.ts]: Don't need a `FileMap`, just a `Map`. Then you can iterate using `forEach`.

File: src/server/project.ts [hunk 10]
```diff
@@ -310,6 +311,13 @@ namespace ts.server {
             this.lsHost.dispose();
             this.lsHost = undefined;
 
+            // Clean up file watchers waiting for missing files
+            for (const p of this.missingFilesMap.getKeys()) {
+                this.missingFilesMap.get(p).close();
+                this.missingFilesMap.remove(p);
```
Comment (andy-ms) [inline src/server/project.ts]: What's the point of removing items one-by-one if you just set `this.missingFileMap = undefined;` at the end?

File: src/server/project.ts [hunk 11]
```diff
@@ -606,6 +614,39 @@ namespace ts.server {
                 }
             }
 
+            if (hasChanges) {
```
Comment (andy-ms) [inline src/server/project.ts]: It was this way before, but I would prefer `const hasChanges = ...;` over `let hasChanges = false; if (...) { hasChanges = true; }

File: src/server/project.ts [hunk 12]
```diff
@@ -606,6 +614,39 @@ namespace ts.server {
                 }
             }
 
+            if (hasChanges) {
+                const missingFilePaths = this.program.getMissingFilePaths() || emptyArray;
```
Comment (andy-ms) [inline src/server/project.ts]: Again, `getMissingFilePaths` should always be defined, so no need for `|| emptyArray`.

File: src/server/project.ts [hunk 13]
```diff
@@ -606,6 +614,39 @@ namespace ts.server {
                 }
             }
 
+            if (hasChanges) {
+                const missingFilePaths = this.program.getMissingFilePaths() || emptyArray;
+                const missingFilePathsSet = createMap<true>();
```
Comment (andy-ms) [inline src/server/project.ts]: `arrayToMap(missingFilePaths, path => path);`

File: src/services/documentHighlights.ts [hunk 14]
```diff
@@ -2,7 +2,10 @@
 namespace ts.DocumentHighlights {
     export function getDocumentHighlights(program: Program, cancellationToken: CancellationToken, sourceFile: SourceFile, position: number, sourceFilesToSearch: SourceFile[]): DocumentHighlights[] | undefined {
         const node = getTouchingWord(sourceFile, position, /*includeJsDocComment*/ true);
-        if (!node) return undefined;
+        // Note that getTouchingWord indicates failure by returning the sourceFile node.
```
Comment (andy-ms) [inline src/services/documentHighlights.ts]: Could this be its own PR?

File: src/server/project.ts [hunk 15]
```diff
@@ -606,6 +614,39 @@ namespace ts.server {
...
+                // should no longer be watched.
+                this.missingFilesMap.getKeys().forEach(p => {
+                    if (!missingFilePathsSet.has(p)) {
+                        this.missingFilesMap.get(p).close();
+                        this.missingFilesMap.remove(p);
+                    }
+                });
+
+                // Missing files that are not yet watched should be added to the map.
+                missingFilePaths.forEach(p => {
```
Comment (andy-ms) [inline src/server/project.ts]: Use a memorable variable name for `p`, like `missingPath`.

File: src/services/documentHighlights.ts (hunk 14 above)
Comment (amcasey) [inline src/services/documentHighlights.ts]: It's a separate commit, so it would be no extra work to make it a separate PR.  What's the local convention?

File: src/compiler/program.ts (hunk 3 above)
Comment (amcasey) [inline src/compiler/program.ts]: I'm happy to update the type, but I should mention that the map already had `undefined` values (this change just ensures that they persist across structure reuse).

File: src/compiler/tsc.ts (hunk 5 above)
Comment (amcasey) [inline src/compiler/tsc.ts]: I was under the impression that we didn't control the signature, but I'll confirm.

File: src/compiler/types.ts (hunk 6 above)
Comment (amcasey) [inline src/compiler/types.ts]: Will do.

File: src/harness/unittests/programMissingFiles.ts (hunk 7 above)
Comment (amcasey) [inline src/harness/unittests/programMissingFiles.ts]: I believe I have a test that mixes existent and non-existent paths, but I'll confirm.

File: src/server/project.ts (hunk 9 above)
Comment (amcasey) [inline src/server/project.ts]: Don't I need a `FileMap` if the key type is `Path`?

File: src/server/project.ts (hunk 10 above)
Comment (amcasey) [inline src/server/project.ts]: Good point.  I cleared the map as an afterthought.

File: src/server/project.ts (hunk 9 above)
Comment (andy-ms) [inline src/server/project.ts]: A `Path` is a subtype of `string`, so you can use those as map keys. (It might be slightly more type safe to specify that keys must be `Path`s, but I don't think that's worth using a `FileMap`.)

File: src/services/documentHighlights.ts (hunk 14 above)
Comment (andy-ms) [inline src/services/documentHighlights.ts]: Ask around?

File: src/compiler/program.ts (hunk 3 above)
Comment (amcasey) [inline src/compiler/program.ts]: To confirm, you're suggesting `const filesByName = createFileMap<SourceFile | undefined>();` ?

Comment (andy-ms) [inline src/compiler/program.ts]: :+1:

File: src/compiler/sys.ts (hunk 4 above)
Comment (amcasey) [inline src/compiler/sys.ts]: I would be interested to learn how to do this.  Do I just specify that it has to have an `mtime` property or is there something more?

File: src/compiler/tsc.ts (hunk 5 above)
Comment (amcasey) [inline src/compiler/tsc.ts]: Done

File: src/compiler/program.ts (hunk 3 above)
Comment (amcasey) [inline src/compiler/program.ts]: That doesn't appear to have any effect on the inferred type.  Presumably, it would matter with strict null-checking enabled?

File: src/harness/unittests/programMissingFiles.ts (hunk 7 above)
Comment (amcasey) [inline src/harness/unittests/programMissingFiles.ts]: I do have such a test, but it is for root files.  I thought that duplicating all of the tests for triple-slash references would be redundant (and add to maintenance and runtime costs), so I skipped them.  Let me know if you feel strongly and I'll make a second set of tests.

File: src/compiler/program.ts (hunk 3 above)
Comment (andy-ms) [inline src/compiler/program.ts]: Yes, it's just documentation for now.

File: src/harness/unittests/programMissingFiles.ts (hunk 7 above)
Comment (andy-ms) [inline src/harness/unittests/programMissingFiles.ts]: No, should be fine then.

File: src/server/project.ts (hunk 13 above)
Comment (amcasey) [inline src/server/project.ts]: Does that have an advantage other than brevity?

Comment (andy-ms) [inline src/server/project.ts]: No, it should have an identical result.

File: src/server/project.ts (hunk 15 above)
Comment (amcasey) [inline src/server/project.ts]: I'm not in the habit of giving detailed names to iteration variables, but sure.

File: src/server/project.ts (hunk 13 above)
Comment (amcasey) [inline src/server/project.ts]: It seems to result in a `String-Path` map, rather than a `Path-true` map.

File: src/server/project.ts (hunk 9 above)
Comment (amcasey) [inline src/server/project.ts]: I don't understand when we use `Path` and `FileMap`.  Do you have some time to explain?

Comment (amcasey) [inline src/server/project.ts]: Also, I've made the change.

File: src/compiler/sys.ts (hunk 4 above)
Comment (amcasey) [inline src/compiler/sys.ts]: ❓

File: src/server/project.ts (hunk 13 above)
Comment (amcasey) [inline src/server/project.ts]: ❓

**Comment by amcasey** [comment]
@andy-ms I think I've addressed all of your feedback, other than the ones marked with :question:.

File: src/server/project.ts (hunk 13 above)
Comment (andy-ms) [inline src/server/project.ts]: There should probably be a helper for this:
```ts
export function arrayToSet(array: string[]): Map<boolean> {
    return arrayToMap(array, key => key, () => true);
}
```

File: src/compiler/sys.ts (hunk 4 above)
Comment (amcasey) [inline src/compiler/sys.ts]: After getting offline help, I don't think this is worth fixing as part of this change.

File: src/services/documentHighlights.ts (hunk 14 above)
Comment (amcasey) [inline src/services/documentHighlights.ts]: I didn't find strong support for pulling it out.  If it were more contentious, I would.

File: src/compiler/core.ts [hunk 16]
```diff
@@ -1101,6 +1101,19 @@ namespace ts {
         return result;
     }
 
+    /**
+     * Creates a set from the elements of an array.
+     *
+     * @param array the array of input elements.
+     *
+     * This function makes no effort to avoid collisions; if any two elements produce
```
Comment (andy-ms) [inline src/compiler/core.ts]: This is a `Map<true>`, so the elements of the array aren't present in the result. So `the element with the higher index in the array will be the one associated with the produced key` can't be true.

File: src/server/project.ts [hunk 17]
```diff
@@ -616,22 +616,24 @@ namespace ts.server {
 
                 // Files that are no longer missing (e.g. because they are no longer required)
                 // should no longer be watched.
-                for (const missingFilePath of this.missingFilesMap.getKeys())
+                const missingFilesToDelete: string[] = [];
```
Comment (andy-ms) [inline src/server/project.ts]: It should be safe to delete values from a map while iterating over it. Are you sure this was the cause of an error?

Comment (amcasey) [inline src/server/project.ts]: No, I just couldn't find any documentation indicating that it was safe.

File: src/compiler/core.ts (hunk 16 above)
Comment (amcasey) [inline src/compiler/core.ts]: Deleted.

**Comment by amcasey** [comment]
After a high-level review from @mhegazy and a detailed review from @andy-ms, I think this is ready to be merged.  Any final comments? (cc: @rbuckton)

File: src/compiler/program.ts [hunk 18]
```diff
@@ -513,6 +513,8 @@ namespace ts {
             }
         }
 
+        const missingFilePaths = filesByName.getKeys().filter(p => !filesByName.get(p));
```
Comment (andy-ms) [inline src/compiler/program.ts]: This will have a semantic conflict with #16724 as `filesByName` will be a `Map`, which doesn't have a `getKeys()`. You can use `mapDefined<[string, string], string>(arrayFrom(m.entries()), ([path, value]) => value === undefined ? undefined : path);`. Or add a version of `mapDefined` that works on iterators to avoid the `arrayFrom`.

File: src/compiler/program.ts [hunk 19]
```diff
@@ -862,6 +865,23 @@ namespace ts {
...
             }
 
+            // If a file has ceased to be missing, then we need to discard some of the old
+            // structure in order to pick it up.
+            // Caution: if the file has created and then deleted between since it was discovered to
+            // be missing, then the corresponding file watcher will have been closed and no new one
+            // will be created until we encounter a change that prevents complete structure reuse.
+            // During this interval, creation of the file will go unnoticed.  We expect this to be
+            // both rare and low-impact.
+            for (const missingFilePath of oldProgram.getMissingFilePaths()) {
```
Comment (andy-ms) [inline src/compiler/program.ts]: Nit: just use `oldProgram.getMissingFilePaths().some(path => host.fileExists(path))`

File: src/compiler/sys.ts [hunk 20]
```diff
@@ -340,11 +346,22 @@ namespace ts {
                     }
 
                     function fileChanged(curr: any, prev: any) {
-                        if (+curr.mtime <= +prev.mtime) {
+                        const isCurrZero = +curr.mtime === 0;
+                        const isPrevZero = +prev.mtime === 0;
+                        const created = !isCurrZero && isPrevZero;
+                        const deleted = isCurrZero && !isPrevZero;
+
+                        const eventKind = created
```
Comment (andy-ms) [inline src/compiler/sys.ts]: Nit: No need for `created` and `deleted` variables, just use `eventKind === FileWatcherEventKind.Changed` instead of `!created && !deleted`

File: src/compiler/tsc.ts [hunk 21]
```diff
@@ -285,6 +285,16 @@ namespace ts {
 
             setCachedProgram(compileResult.program);
             reportWatchDiagnostic(createCompilerDiagnostic(Diagnostics.Compilation_complete_Watching_for_file_changes));
+
+            const missingPaths = compileResult.program.getMissingFilePaths();
+            missingPaths.forEach((path: Path): void => {
```
Comment (andy-ms) [inline src/compiler/tsc.ts]: Nit: Avoid annotating the types of callbacks if not necessary.

File: src/harness/unittests/programMissingFiles.ts (hunk 8 above)
Comment (andy-ms) [inline src/harness/unittests/programMissingFiles.ts]: No need for `assert.isDefined` since you're about to assert that it's equal to a defined value.

File: src/server/server.ts [hunk 22]
```diff
@@ -522,13 +522,19 @@ namespace ts.server {
                 return;
             }
 
+            // removed = deleted ? true : (added ? false : undefined)
```
Comment (andy-ms) [inline src/server/server.ts]: comment outdated

File: src/services/documentHighlights.ts (hunk 14 above)
Comment (andy-ms) [inline src/services/documentHighlights.ts]: Was there a test that was breaking under the old behavior?

Comment (amcasey) [inline src/services/documentHighlights.ts]: No, otherwise the code would have been correct. :wink:

File: src/compiler/program.ts (hunk 18 above)
Comment (amcasey) [inline src/compiler/program.ts]: I'll rebase on top of your commit.

File: src/compiler/sys.ts (hunk 20 above)
Comment (amcasey) [inline src/compiler/sys.ts]: I think the existence of the constants makes the code more readable, but I agree that the `if` is better expressed in terms of `eventKind`.

File: src/server/server.ts (hunk 22 above)
Comment (amcasey) [inline src/server/server.ts]: Oops, missed one.

File: src/server/project.ts [hunk 23]
```diff
@@ -604,6 +609,35 @@ namespace ts.server {
...
+                    if (!missingFilePathsSet.has(missingFilePath)) {
+                        this.missingFilesMap.delete(missingFilePath);
+                        fileWatcher.close();
+                    }
+                });
+
+                // Missing files that are not yet watched should be added to the map.
+                for (const missingFilePath of missingFilePaths) {
+                    if (!this.missingFilesMap.has(missingFilePath)) {
+                        const fileWatcher = this.projectService.host.watchFile(missingFilePath, (_filename: string, eventKind: FileWatcherEventKind) => {
```
Comment (andy-ms) [inline src/server/project.ts]: `host.watchFile` is actually optional. We could make it non-optional and mark it as a breaking change. See also #16776.

File: src/compiler/core.ts [hunk 24]
```diff
@@ -1097,6 +1097,15 @@ namespace ts {
         return result;
     }
 
+    /**
+     * Creates a set from the elements of an array.
+     *
+     * @param array the array of input elements.
+     */
+    export function arrayToSet<T>(array: T[], makeKey: (value: T) => string): Map<true> {
```
Comment (andy-ms) [inline src/compiler/core.ts]: It's generally useful to factor generic code out of a more specific context. That way you don't have to worry about the specifics of how an array is converted to a set, and only have to worry about the actual problem domain.
